File: src/layout/dagre/src/order/sort-subgraph.ts
import { Graph } from "../../graph";
import barycenter from "./barycenter";
import resolveConflicts, { ConflictEntry } from "./resolve-conflicts";
import sort from "./sort";

const sortSubgraph = (
  g: Graph,
  v: string,
  cg: Graph,
  biasRight?: boolean,
  usePrev?: boolean
) => {
  let movable = g.children(v);
  // fixorder的点不参与排序（这个方案不合适，只排了新增节点，和原来的分离）
  const node = g.node(v)!;
  const bl = node ? node.borderLeft : undefined;
  const br = node ? node.borderRight : undefined;
  const subgraphs: Record<string, Partial<ConflictEntry>> = {};

  if (bl) {
    movable = movable?.filter((w) => {
      return w !== bl && w !== br;
    });
  }

  const barycenters = barycenter(g, movable || []);
  barycenters?.forEach((entry) => {
    if (g.children(entry.v)?.length) {
      const subgraphResult = sortSubgraph(g, entry.v, cg, biasRight);
      subgraphs[entry.v] = subgraphResult;
      if (subgraphResult.hasOwnProperty("barycenter")) {
        mergeBarycenters(entry, subgraphResult);
      }
    }
  });

  const entries = resolveConflicts(barycenters, cg);
  expandSubgraphs(entries, subgraphs);

  // 添加fixorder信息到entries里边
  // TODO: 不考虑复合情况，只用第一个点的fixorder信息，后续考虑更完备的实现
  entries
    .filter((e) => e.vs.length > 0)
    ?.forEach((e) => {
      const node = g.node(e.vs[0])!;
      if (node) {
        e.fixorder = node.fixorder;
        e.order = node.order;
      }
    });

  const result = sort(entries, biasRight, usePrev);

  if (bl) {
    result.vs = [bl, result.vs, br].flat();
    if (g.predecessors(bl)?.length) {
      const blPred = g.node(g.predecessors(bl)?.[0] || "")!;
      const brPred = g.node(g.predecessors(br)?.[0] || "")!;
      if (!result.hasOwnProperty("barycenter")) {
        result.barycenter = 0;
        result.weight = 0;
      }
      result.barycenter =
        (result.barycenter! * result.weight! +
          (blPred.order as number) +
          (brPred.order as number)) /
        (result.weight! + 2);
      result.weight! += 2;
    }
  }

  return result;
};

const expandSubgraphs = (
  entries: ConflictEntry[],
  subgraphs: Record<string, Partial<ConflictEntry>>
) => {
  entries?.forEach((entry) => {
    const vss = entry.vs?.map((v: string) => {
      if (subgraphs[v]) {
        return subgraphs[v].vs!;
      }
      return v;
    });
    entry.vs = vss.flat();
  });
};

const mergeBarycenters = (
  target: { weight?: number; barycenter?: number },
  other: { weight?: number; barycenter?: number }
) => {
  if (target.barycenter !== undefined) {
    target.barycenter =
      (target.barycenter * target.weight! + other.barycenter! * other.weight!) /
      (target.weight! + other.weight!);
    target.weight! += other.weight!;
  } else {
    target.barycenter = other.barycenter;
    target.weight = other.weight;
  }
};

export default sortSubgraph;

File: packages/layout-rust/src/dagre/order/sort_subgraph.rs
use crate::dagre::order::barycenter::{barycenter, Barycenter};
use crate::dagre::order::resolve_conflicts::{resolve_conflicts, ResolvedBaryEntry};
use crate::dagre::order::sort::sort;
use crate::{GraphConfig, GraphEdge, GraphNode};
use graphlib_rust::Graph;
use ordered_hashmap::OrderedHashMap;

#[derive(Debug, Clone, Default)]
pub struct SubgraphResult {
    pub vs: Vec<String>,
    pub barycenter: f32,
    pub weight: f32,
}

pub fn sort_subgraph(
    g: &Graph<GraphConfig, GraphNode, GraphEdge>,
    v: &String,
    cg: &Graph<GraphConfig, GraphNode, GraphEdge>,
    bias_right: &bool,
) -> SubgraphResult {
    let mut movable = g.children(v);
    let node = g.node(v);
    let bl = if node.is_some() {
        node.unwrap().border_left_.clone()
    } else {
        None
    };
    let br = if node.is_some() {
        node.unwrap().border_right_.clone()
    } else {
        None
    };
    let mut subgraphs: OrderedHashMap<String, SubgraphResult> = OrderedHashMap::new();

    if br.is_some() {
        movable = movable
            .into_iter()
            .filter(|w| w != &bl.clone().unwrap() && w != &br.clone().unwrap())
            .collect();
    }

    let mut barycenters = barycenter(g, &movable);
    barycenters.iter_mut().for_each(|entry| {
        if g.children(&entry.v).len() > 0 {
            let subgraph_result = sort_subgraph(g, &entry.v, cg, bias_right);
            subgraphs.insert(entry.v.clone(), subgraph_result.clone());
            merge_barycenters(entry, &subgraph_result);
        }
    });

    let mut entries = resolve_conflicts(&barycenters, cg);
    expand_subgraphs(&mut entries, &subgraphs);

    let mut result = sort(&entries, bias_right);
    if bl.is_some() {
        let bl_ = bl.clone().unwrap();
        let br_ = br.clone().unwrap();
        result.vs = vec![vec![bl_.clone()], result.vs.clone(), vec![br_.clone()]].concat();
        let bl_preds = g.predecessors(&bl_).unwrap_or(vec![]);
        if bl_preds.len() > 0 {
            let bl_pred = g.node(bl_preds.first().unwrap()).unwrap();
            let br_pred = g
                .node(g.predecessors(&br_).unwrap_or(vec![]).first().unwrap())
                .unwrap();
            // not required as it was handled by structs' default
            /*
            if (!_.has(result, "barycenter")) {
              result.barycenter = 0;
              result.weight = 0;
            }
            */
            let bl_pred_order = bl_pred.order.clone().unwrap_or(0) as f32;
            let br_pred_order = br_pred.order.clone().unwrap_or(0) as f32;
            result.barycenter = (result.barycenter * result.weight + bl_pred_order + br_pred_order)
                / (result.weight + 2.0);
            result.weight += 2.0;
        }
    }

    return result;
}

fn expand_subgraphs(
    entries: &mut Vec<ResolvedBaryEntry>,
    subgraphs: &OrderedHashMap<String, SubgraphResult>,
) {
    entries.iter_mut().for_each(|entry| {
        let mut vs: Vec<String> = vec![];
        entry.vs.iter().for_each(|v| {
            if subgraphs.contains_key(v) {
                let subgraph = subgraphs.get(v).unwrap();
                subgraph.vs.iter().for_each(|v_| {
                    vs.push(v_.clone());
                });
                return ();
            }
            vs.push(v.clone());
        });

        entry.vs = vs;
    });
}

fn merge_barycenters(target: &mut Barycenter, other: &SubgraphResult) {
    let target_barycenter = target.barycenter.clone().unwrap_or(0.0);
    if target.barycenter.is_some() && target_barycenter > 0.0 {
        let target_weight = target.weight.clone().unwrap_or(0.0);

        target.barycenter = Some(
            (target_barycenter * target_weight + other.barycenter * other.weight)
                / (target_weight + other.weight),
        );
        target.weight = Some(target_weight + other.weight.clone());
    } else {
        target.barycenter = Some(other.barycenter.clone());
        target.weight = Some(other.weight.clone());
    }
}
